storage/local/frankenstein.go

```diff
@@ -310,32 +371,31 @@ func (i *Ingestor) flushSeries(fp model.Fingerprint, series *memorySeries, immed
 	if !series.headChunkClosed {
 		chunks = chunks[:len(chunks)-1]
 	}
+	u.fpLocker.Unlock(fp)
 	if len(chunks) == 0 {
-		i.fpLocker.Unlock(fp)
 		return nil
 	}
 
 	// flush the chunks without locking the series
-	i.fpLocker.Unlock(fp)
 	log.Infof("Flushing %d chunks", len(chunks))
-	if err := i.flushChunks(fp, series.metric, chunks); err != nil {
+	if err := i.flushChunks(ctx, fp, series.metric, chunks); err != nil {
 		i.chunkStoreFailures.Add(float64(len(chunks)))
 		return err
 	}
 	i.storedChunks.Add(float64(len(chunks)))
 
 	// now remove the chunks
-	i.fpLocker.Lock(fp)
+	u.fpLocker.Lock(fp)
 	series.chunkDescs = series.chunkDescs[len(chunks):]
 	if len(series.chunkDescs) == 0 {
-		i.fpToSeries.del(fp)
-		i.index.delete(series.metric, fp)
+		u.fpToSeries.del(fp)
+		u.index.delete(series.metric, fp)
```

Comment:
Add a TODO to delete an entire entry from the user state map if the last series for a user was removed (or implement it :)).